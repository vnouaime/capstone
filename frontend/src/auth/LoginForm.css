.loginform-header {
    margin-top: 30px;
}

.loginform-card {
    border-color: rgb(137, 225, 169);
}

.loginform-input {
    border-color: rgb(137, 225, 169);
}

.loginform-button {
    border-color:rgb(137, 225, 169);
    background-color: rgb(137, 225, 169);
    color: white;
}