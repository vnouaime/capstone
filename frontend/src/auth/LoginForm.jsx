import React, { useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import { Form, Button, Container, Card } from "react-bootstrap";

import Alert from "../common/Alert";

import "./LoginForm.css"

/** LoginForm
 *
 * Renders form for logging in existing user. 
 * Changes state as user types in form. 
 * When submitted, calls login function passed from prop. 
 * If successful, redirects to /. If not, 
 * alerts user with message. 
 * 
 * Props: 
 * login
 * 
 * State: 
 * formData, formErrors
 *
 * Routes -> LoginForm -> Alert
 * Routed as /login
 */
const LoginForm = ({ login }) => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  async function handleSubmit(evt) {
    evt.preventDefault();

    let result = await login(formData);
    if (result.success) {
      navigate("/");
    } else {
      setFormErrors(result.errors);
    }
  }

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((data) => ({ ...data, [name]: value }));
  }

  return (
    <Container>
      <Container className="col-md-10 offset-md-1 col-lg-8 offset-lg-2 mb-5">
        <h1 className="loginform-header mb-3 text-center">Login</h1>
        {formErrors.length ? (
                <Alert type="danger" messages={formErrors} />
              ) : null}
        <Card>
          <Card.Body>
            <Form onSubmit={handleSubmit}>
              <Form.Group className="mb-3">
                <Form.Label htmlFor="username">Username</Form.Label>
                <Form.Control
                    className="loginform-input"
                    id="username"
                    name="username"
                    type="text"
                    placeholder="Username"
                    value={formData.username}
                    onChange={handleChange}
                    autoComplete="username"
                    required
                />
            </Form.Group>
            <Form.Group className="mb-3">
                <Form.Label htmlFor="password">Password</Form.Label>
                <Form.Control
                    className="loginform-input"
                    id="password"
                    type="password"
                    name="password"
                    placeholder="Password"
                    value={formData.password}
                    onChange={handleChange}
                    autoComplete="current-password"
                    required
                />
            </Form.Group>
              <div className="d-flex justify-content-end">
                <Button className="loginform-button" type="submit" data-testid="login-button">
                  Login
                </Button>
              </div>
            </Form>
          </Card.Body>
        </Card>
      </Container>
      <div className="text-center">Don't have an account? <Link to="/signup">Sign up here</Link></div>
    </Container>
  );
}

export default LoginForm;
