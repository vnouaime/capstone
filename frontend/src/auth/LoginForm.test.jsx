import React from "react";
import { expect, vi } from "vitest"
import { render, fireEvent } from "vitest-react";
import { MemoryRouter } from "react-router-dom";
import LoginForm from "./LoginForm";

test("smoke test", () => {
    render(
        <MemoryRouter>
            <LoginForm />
        </MemoryRouter>
    );
});

test("snapshot", () => {
    const { asFragment } = render(<MemoryRouter><LoginForm /></MemoryRouter>)
    expect(asFragment()).toMatchSnapshot()
})

test("should login user", async () => {
    const login = vi.fn().mockResolvedValue({ success: true })

    const { getByTestId, getByLabelText } = 
        render(
            <MemoryRouter>
                <LoginForm login={login}/>
            </MemoryRouter>)

    const username = getByLabelText("Username")
    const password = getByLabelText("Password")
    const loginButton = getByTestId("login-button")
    
    fireEvent.change(username, { target: { value: 'testuser' } })
    fireEvent.change(password, { target: { value: 'password' } })
    fireEvent.click(loginButton)

    await vi.waitFor(() => expect(window.location.pathname).toBe('/'));
})

test("unauthorized login", async () => {
    const login = vi.fn().mockResolvedValue({ success: false, errors: ["Invalid username/password"]})

    const { getByTestId, getByLabelText, findByText } = 
        render(
            <MemoryRouter>
                <LoginForm login={login}/>
            </MemoryRouter>)

    const username = getByLabelText("Username")
    const password = getByLabelText("Password")
    const loginButton = getByTestId("login-button")
    
    fireEvent.change(username, { target: { value: 'testuser' } })
    fireEvent.change(password, { target: { value: 'password' } })
    fireEvent.click(loginButton)

    const alert = await findByText("Invalid username/password");
    expect(alert).toBeInTheDocument();
    expect(alert.closest('div')).toHaveClass("alert-danger");
})
