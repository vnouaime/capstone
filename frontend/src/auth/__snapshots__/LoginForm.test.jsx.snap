// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="col-md-10 offset-md-1 col-lg-8 offset-lg-2 mb-5 container"
    >
      <h1
        class="loginform-header mb-3 text-center"
      >
        Login
      </h1>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form
            class=""
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="username"
                class="loginform-input form-control"
                id="username"
                name="username"
                placeholder="Username"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="loginform-input form-control"
                id="password"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
            </div>
            <div
              class="d-flex justify-content-end"
            >
              <button
                class="loginform-button btn btn-primary"
                data-testid="login-button"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="text-center"
    >
      Don't have an account? 
      <a
        href="/signup"
      >
        Sign up here
      </a>
    </div>
  </div>
</DocumentFragment>
`;
